cli/commands/catalog/module/repo.go

```diff
@@ -0,0 +1,257 @@
+func moduleDocPath(modulePath string) string {
+	if !files.FileExists(filepath.Join(modulePath, "main.tf")) || !files.FileExists(filepath.Join(modulePath, "variables.tf")) {
+		return ""
+	}
+
+	for _, filename := range docFilenames {
+		path := filepath.Join(modulePath, filename)
+		if files.FileExists(path) {
+			return path
+		}
+	}
+
+	return ""
+}
+
+// module returns a module instance if the given path `repoPath/moduleDir` contains a Terragrunt module.
+func module(repoName, repoPath, moduleDir string) (*Module, error) {
+	var (
+		modulePath = filepath.Join(repoPath, moduleDir)
+
+		reg       = mdHeaderReg
+		docHeader = mdHeader
+
+		title            = filepath.Base(moduleDir)
+		descriptionLines []string
+	)
+
+	docPath := moduleDocPath(modulePath)
+	if docPath == "" {
+		return nil, nil
+	}
+
+	log.Debugf("Found Terragrunt module in directory %q", modulePath)
+
+	docContentByte, err := os.ReadFile(docPath)
+	if err != nil {
+		return nil, errors.WithStackTrace(err)
+	}
+	docContent := string(docContentByte)
+
+	if strings.HasSuffix(docPath, ".adoc") {
+		reg = adocHeaderReg
+		docHeader = adocHeader
+	}
+
+	if match := reg.FindStringSubmatch(docContent); len(match) > 0 {
+		header := match[1]
+
+		// remove comments
+		header = commentReg.ReplaceAllString(header, "")
+		// remove adoc images
+		header = adocImageReg.ReplaceAllString(header, "")
+
+		lines := strings.Split(header, "\n")
+		title = strings.TrimSpace(lines[0])
+
+		if len(lines) > 1 {
+			for _, line := range lines[1:] {
+				line = strings.TrimSpace(line)
+
+				// another header begins
+				if strings.HasPrefix(line, docHeader) {
+					break
+				}
+
+				descriptionLines = append(descriptionLines, line)
+			}
+		}
+	}
+
+	return &Module{
+		path:        modulePath,
+		url:         path.Join(gitHubOrgURL, repoName, "tree/master", moduleDir),
+		title:       title,
+		description: strings.TrimSpace(strings.Join(descriptionLines, " ")),
+		content:     docContent,
+	}, nil
+
+}
+
+// FindModules clones the repository if `repoPath` is a URL, searches for Terragrunt modules, indexes their README.* files, and returns module instances.
+func FindModules(ctx context.Context, repoPath string) (Modules, error) {
+	var repoName string
+
+	tempDir, err := os.MkdirTemp("", "catalog-*")
+	if err != nil {
+		return nil, errors.WithStackTrace(err)
+	}
+	//nolint:errcheck
+	defer os.RemoveAll(tempDir)
+
+	repoPath, err = getRepo(ctx, repoPath, tempDir)
+	if err != nil {
+		return nil, err
+	}
+
+	remoteURL, err := gitRemoteURL(repoPath)
+	if err != nil {
+		return nil, err
+	}
+	// remove extension like `.git`
+	ext := filepath.Ext(remoteURL)
+	remoteURL = strings.TrimRight(remoteURL, "."+ext)
+
+	repoName = filepath.Base(remoteURL)
+
+	modulesPath := filepath.Join(repoPath, "modules")
+
+	// It is assumed that the repository is a module itself
+	if !files.FileExists(modulesPath) {
```

Comment:
Some repos that have a `modules` folder _also_ have a module in the root of the repo.

Reply:
> Some repos that have a modules folder also have a module in the root of the repo.
An, udnerstood. Fixed.

>Also, one more thought: are we checking all sub-folders of modules? Because in some repos, we have modules one level below (e.g., modules/vpc) and in some, it's two levels (modules/networking/vpc). More importantly, do we have a test for that?

Yes, we recursively test all `module/*` subdirectories.